algorithms/linfa-nn/src/distance.rs

```diff
@@ -3,7 +3,7 @@ use ndarray::{Array2, ArrayBase, ArrayView, Axis, Data, Dimension, Ix2, Zip};
 use ndarray_stats::DeviationExt;
 
 /// A distance function that can be used in spatial algorithms such as nearest neighbour.
-pub trait Distance<F: Float>: Clone + Send + Sync {
+pub trait Distance<F: Float>: Clone + Send + Sync + Unpin + Sized {
```

Comment:
Why add these bounds? The autotrait tests are on concrete types, not generics, so they should pass even without the bounds.

Reply:
Removed `Sized`, but `Unpin` was added due to the bound on `NearestNeighbourIndex`

Reply:
Why do the bounds on `NearestNeighbourIndex` (which are needed due to our use of trait objects) propagate to `Distance`? What error do you get when you remove the `Unpin` bound?

Reply:
Those are all the errors I get when removing `Unpin` bound from `Distance`
```
error[E0277]: `D` cannot be unpinned
   --> algorithms/linfa-nn/src/balltree.rs:263:36
    |
263 | impl<'a, F: Float, D: Distance<F>> NearestNeighbourIndex<F> for BallTreeIndex<'a, F, D> {
    |                                    ^^^^^^^^^^^^^^^^^^^^^^^^ within `BallTreeIndex<'a, F, D>`, the trait `Unpin` is not implemented for `D`
    |
    = note: consider using `Box::pin`
note: required because it appears within the type `BallTreeIndex<'a, F, D>`
   --> algorithms/linfa-nn/src/balltree.rs:163:12
    |
163 | pub struct BallTreeIndex<'a, F: Float, D: Distance<F>> {
    |            ^^^^^^^^^^^^^
note: required by a bound in `NearestNeighbourIndex`
   --> algorithms/linfa-nn/src/lib.rs:90:58
    |
90  | pub trait NearestNeighbourIndex<F: Float>: Send + Sync + Unpin {
    |                                                          ^^^^^ required by this bound in `NearestNeighbourIndex`
help: consider further restricting this bound
    |
263 | impl<'a, F: Float, D: Distance<F> + std::marker::Unpin> NearestNeighbourIndex<F> for BallTreeIndex<'a, F, D> {
    |                                   ++++++++++++++++++++

error[E0277]: `D` cannot be unpinned
  --> algorithms/linfa-nn/src/kdtree.rs:52:36
   |
52 | impl<'a, F: Float, D: Distance<F>> NearestNeighbourIndex<F> for KdTreeIndex<'a, F, D> {
   |                                    ^^^^^^^^^^^^^^^^^^^^^^^^ within `KdTreeIndex<'a, F, D>`, the trait `Unpin` is not implemented for `D`
   |
   = note: consider using `Box::pin`
note: required because it appears within the type `KdTreeIndex<'a, F, D>`
  --> algorithms/linfa-nn/src/kdtree.rs:12:12
   |
12 | pub struct KdTreeIndex<'a, F: Float, D: Distance<F>>(
   |            ^^^^^^^^^^^
note: required by a bound in `NearestNeighbourIndex`
  --> algorithms/linfa-nn/src/lib.rs:90:58
   |
90 | pub trait NearestNeighbourIndex<F: Float>: Send + Sync + Unpin {
   |                                                          ^^^^^ required by this bound in `NearestNeighbourIndex`
help: consider further restricting this bound
   |
52 | impl<'a, F: Float, D: Distance<F> + std::marker::Unpin> NearestNeighbourIndex<F> for KdTreeIndex<'a, F, D> {
   |                                   ++++++++++++++++++++

For more information about this error, try `rustc --explain E0277`.
error: could not compile `linfa-nn` due to 2 previous errors
mila@mila-ESPRIMO-P957:~/Documents/linfa/algorithms/linfa-clustering$ cargo build
   Compiling linfa-nn v0.5.0 (/home/mila/Documents/linfa/algorithms/linfa-nn)
error[E0277]: `D` cannot be unpinned
   --> algorithms/linfa-nn/src/balltree.rs:263:36
    |
263 | impl<'a, F: Float, D: Distance<F>> NearestNeighbourIndex<F> for BallTreeIndex<'a, F, D> {
    |                                    ^^^^^^^^^^^^^^^^^^^^^^^^ within `BallTreeIndex<'a, F, D>`, the trait `Unpin` is not implemented for `D`
    |
    = note: consider using `Box::pin`
note: required because it appears within the type `BallTreeIndex<'a, F, D>`
   --> algorithms/linfa-nn/src/balltree.rs:163:12
    |
163 | pub struct BallTreeIndex<'a, F: Float, D: Distance<F>> {
    |            ^^^^^^^^^^^^^
note: required by a bound in `NearestNeighbourIndex`
   --> algorithms/linfa-nn/src/lib.rs:90:58
    |
90  | pub trait NearestNeighbourIndex<F: Float>: Send + Sync + Unpin {
    |                                                          ^^^^^ required by this bound in `NearestNeighbourIndex`
help: consider further restricting this bound
    |
263 | impl<'a, F: Float, D: Distance<F> + std::marker::Unpin> NearestNeighbourIndex<F> for BallTreeIndex<'a, F, D> {
    |                                   ++++++++++++++++++++

error[E0277]: `D` cannot be unpinned
  --> algorithms/linfa-nn/src/kdtree.rs:52:36
   |
52 | impl<'a, F: Float, D: Distance<F>> NearestNeighbourIndex<F> for KdTreeIndex<'a, F, D> {
   |                                    ^^^^^^^^^^^^^^^^^^^^^^^^ within `KdTreeIndex<'a, F, D>`, the trait `Unpin` is not implemented for `D`
   |
   = note: consider using `Box::pin`
note: required because it appears within the type `KdTreeIndex<'a, F, D>`
  --> algorithms/linfa-nn/src/kdtree.rs:12:12
   |
12 | pub struct KdTreeIndex<'a, F: Float, D: Distance<F>>(
   |            ^^^^^^^^^^^
note: required by a bound in `NearestNeighbourIndex`
  --> algorithms/linfa-nn/src/lib.rs:90:58
   |
90 | pub trait NearestNeighbourIndex<F: Float>: Send + Sync + Unpin {
   |                                                          ^^^^^ required by this bound in `NearestNeighbourIndex`
help: consider further restricting this bound
   |
52 | impl<'a, F: Float, D: Distance<F> + std::marker::Unpin> NearestNeighbourIndex<F> for KdTreeIndex<'a, F, D> {
   |                                   ++++++++++++++++++++

error[E0277]: `D` cannot be unpinned
  --> algorithms/linfa-nn/src/linear.rs:32:36
   |
32 | impl<'a, F: Float, D: Distance<F>> NearestNeighbourIndex<F> for LinearSearchIndex<'a, F, D> {
   |                                    ^^^^^^^^^^^^^^^^^^^^^^^^ within `LinearSearchIndex<'a, F, D>`, the trait `Unpin` is not implemented for `D`
   |
   = note: consider using `Box::pin`
note: required because it appears within the type `LinearSearchIndex<'a, F, D>`
  --> algorithms/linfa-nn/src/linear.rs:16:12
   |
16 | pub struct LinearSearchIndex<'a, F: Float, D: Distance<F>>(ArrayView2<'a, F>, D);
   |            ^^^^^^^^^^^^^^^^^
note: required by a bound in `NearestNeighbourIndex`
  --> algorithms/linfa-nn/src/lib.rs:90:58
   |
90 | pub trait NearestNeighbourIndex<F: Float>: Send + Sync + Unpin {
   |                                                          ^^^^^ required by this bound in `NearestNeighbourIndex`
help: consider further restricting this bound
   |
32 | impl<'a, F: Float, D: Distance<F> + std::marker::Unpin> NearestNeighbourIndex<F> for LinearSearchIndex<'a, F, D> {
   |                                   ++++++++++++++++++++

For more information about this error, try `rustc --explain E0277`.
error: could not compile `linfa-nn` due to 3 previous errors
```

Reply:
OK so you need the `Unpin` bound because `Distance` is used by the NN traits, which also have `Unpin` bounds. Then keep it as is.